File: rust/src/adapters/irc.rs
//! IRC adapter using the irc crate.

use async_trait::async_trait;
use chrono::Utc;
use std::sync::Arc;
use tokio::sync::Mutex;

use irc::client::prelude::*;

use crate::adapter::{ChannelAdapter, MessageHandler};
use crate::error::{Error, Result};
use crate::types::*;

/// IRC adapter configuration.
pub struct IrcConfig {
    /// IRC server hostname.
    pub server: String,
    /// IRC server port (default 6697 for TLS).
    pub port: u16,
    /// Use TLS.
    pub use_tls: bool,
    /// Nickname.
    pub nickname: String,
    /// Username (ident).
    pub username: Option<String>,
    /// Real name / GECOS.
    pub realname: Option<String>,
    /// Channels to join on connect.
    pub channels: Vec<String>,
    /// NickServ password for identification.
    pub nickserv_password: Option<String>,
}

impl IrcConfig {
    /// Create with minimal required fields.
    pub fn new(server: impl Into<String>, nickname: impl Into<String>) -> Self {
        Self {
            server: server.into(),
            port: 6697,
            use_tls: true,
            nickname: nickname.into(),
            username: None,
            realname: None,
            channels: Vec::new(),
            nickserv_password: None,
        }
    }

    /// Builder: add a channel to auto-join.
    pub fn with_channel(mut self, channel: impl Into<String>) -> Self {
        self.channels.push(channel.into());
        self
    }
}

/// IRC adapter.
pub struct IrcAdapter {
    config: IrcConfig,
    connected: bool,
    last_activity: Option<chrono::DateTime<Utc>>,
    handler: Arc<Mutex<Option<MessageHandler>>>,
    client: Option<Arc<Mutex<irc::client::Client>>>,
    shutdown_tx: Option<tokio::sync::oneshot::Sender<()>>,
}

impl IrcAdapter {
    /// Create a new IRC adapter.
    pub fn new(config: IrcConfig) -> Self {
        Self {
            config,
            connected: false,
            last_activity: None,
            handler: Arc::new(Mutex::new(None)),
            client: None,
            shutdown_tx: None,
        }
    }

    /// Convert an IRC message to a UnifiedMessage.
    fn irc_to_unified(prefix: Option<&str>, target: &str, text: &str) -> UnifiedMessage {
        let nick = prefix
            .and_then(|p| p.split('!').next())
            .unwrap_or("unknown");

        let is_cmd = text.starts_with('!') || text.starts_with('.');
        let content = if is_cmd {
            let stripped = &text[1..];
            let parts: Vec<&str> = stripped.split_whitespace().collect();
            let cmd = parts.first().unwrap_or(&"").to_string();
            let args: Vec<String> = parts[1..].iter().map(|s| s.to_string()).collect();
            MessageContent::command(text, cmd, args)
        } else {
            MessageContent::text(text)
        };

        UnifiedMessage {
            id: uuid::Uuid::new_v4().to_string(),
            channel: "irc".to_string(),
            sender: Identity::new(nick),
            content,
            timestamp: Utc::now(),
            thread_id: None,
            reply_to_id: None,
            chat_id: Some(target.to_string()),
            raw: None,
            metadata: Default::default(),
        }
    }
}

#[async_trait]
impl ChannelAdapter for IrcAdapter {
    fn channel_id(&self) -> &str {
        "irc"
    }

    async fn connect(&mut self) -> Result<()> {
        let irc_config = Config {
            nickname: Some(self.config.nickname.clone()),
            username: self.config.username.clone(),
            realname: self.config.realname.clone(),
            server: Some(self.config.server.clone()),
            port: Some(self.config.port),
            use_tls: Some(self.config.use_tls),
            channels: self.config.channels.clone(),
            nick_password: self.config.nickserv_password.clone(),
            ..Config::default()
        };

        let client = irc::client::Client::from_config(irc_config)
            .await
            .map_err(|e| Error::Connection(format!("IRC connect: {}", e)))?;

        client
            .identify()
            .map_err(|e| Error::Connection(format!("IRC identify: {}", e)))?;

        let client = Arc::new(Mutex::new(client));
        self.client = Some(Arc::clone(&client));

        // Start message listener
        let handler = Arc::clone(&self.handler);
        let (tx, mut rx) = tokio::sync::oneshot::channel::<()>();
        self.shutdown_tx = Some(tx);

        tokio::spawn(async move {
            use futures::StreamExt;
            let mut stream = {
                let c = client.lock().await;
                c.stream().unwrap()
            };

            loop {
                tokio::select! {
                    msg = stream.next() => {
                        match msg {
                            Some(Ok(message)) => {
                                if let Command::PRIVMSG(ref target, ref text) = message.command {
                                    let prefix = message.prefix.as_ref().map(|p| p.to_string());
                                    let unified = IrcAdapter::irc_to_unified(
                                        prefix.as_deref(),
                                        target,
                                        text,
                                    );
                                    let h = handler.lock().await;
                                    if let Some(ref callback) = *h {
                                        callback(unified);
                                    }
                                }
                            }
                            Some(Err(e)) => {
                                tracing::error!(error = %e, "IRC stream error");
                                break;
                            }
                            None => break,
                        }
                    }
                    _ = &mut rx => break,
                }
            }
        });

        self.connected = true;
        Ok(())
    }

    async fn disconnect(&mut self) -> Result<()> {
        if let Some(tx) = self.shutdown_tx.take() {
            let _ = tx.send(());
        }
        if let Some(ref client) = self.client {
            let c = client.lock().await;
            c.send_quit("unified-channel shutdown")
                .map_err(|e| Error::Connection(format!("IRC quit: {}", e)))?;
        }
        self.connected = false;
        Ok(())
    }

    fn on_message(&mut self, handler: MessageHandler) {
        let h = Arc::clone(&self.handler);
        tokio::task::block_in_place(|| {
            let rt = tokio::runtime::Handle::current();
            rt.block_on(async {
                let mut guard = h.lock().await;
                *guard = Some(handler);
            });
        });
    }

    async fn send(&self, msg: OutboundMessage) -> Result<Option<String>> {
        let client = self.client.as_ref().ok_or_else(|| {
            Error::Send("IRC not connected".to_string())
        })?;

        let c = client.lock().await;
        c.send_privmsg(&msg.chat_id, &msg.text)
            .map_err(|e| Error::Send(format!("IRC send: {}", e)))?;

        Ok(None) // IRC doesn't return message IDs
    }

    async fn get_status(&self) -> ChannelStatus {
        ChannelStatus {
            connected: self.connected,
            channel: "irc".to_string(),
            account_id: Some(self.config.nickname.clone()),
            error: None,
            last_activity: self.last_activity,
        }
    }
}

File: python/unified_channel/adapters/irc.py
"""IRC adapter — plain IRC protocol via asyncio sockets.

No external dependencies.
"""

from __future__ import annotations

import asyncio
import logging
import re
from datetime import datetime
from typing import AsyncIterator

from ..adapter import ChannelAdapter
from ..types import (
    ChannelStatus, ContentType, Identity, MessageContent,
    OutboundMessage, UnifiedMessage,
)

logger = logging.getLogger(__name__)

_PRIVMSG_RE = re.compile(r"^:(\S+)!(\S+) PRIVMSG (\S+) :(.+)$")


class IRCAdapter(ChannelAdapter):
    """Plain IRC adapter using asyncio streams."""

    channel_id = "irc"

    def __init__(
        self,
        server: str,
        port: int = 6667,
        nickname: str = "unified-bot",
        channels: list[str] | None = None,
        *,
        password: str | None = None,
        use_ssl: bool = False,
        command_prefix: str = "!",
    ) -> None:
        self._server = server
        self._port = port
        self._nickname = nickname
        self._channels = channels or []
        self._password = password
        self._use_ssl = use_ssl
        self._prefix = command_prefix
        self._queue: asyncio.Queue[UnifiedMessage] = asyncio.Queue()
        self._connected = False
        self._last_activity: datetime | None = None
        self._reader: asyncio.StreamReader | None = None
        self._writer: asyncio.StreamWriter | None = None
        self._listen_task: asyncio.Task | None = None
        self._msg_counter = 0

    async def connect(self) -> None:
        if self._use_ssl:
            import ssl as _ssl
            ctx = _ssl.create_default_context()
            self._reader, self._writer = await asyncio.open_connection(
                self._server, self._port, ssl=ctx,
            )
        else:
            self._reader, self._writer = await asyncio.open_connection(
                self._server, self._port,
            )

        if self._password:
            await self._send_raw(f"PASS {self._password}")
        await self._send_raw(f"NICK {self._nickname}")
        await self._send_raw(f"USER {self._nickname} 0 * :{self._nickname}")

        self._connected = True
        self._listen_task = asyncio.create_task(self._listen())

        # Wait for MOTD end (001 or 376/422)
        await asyncio.sleep(2)
        for ch in self._channels:
            await self._send_raw(f"JOIN {ch}")

        logger.info("irc connected: %s@%s:%d channels=%s",
                     self._nickname, self._server, self._port, self._channels)

    async def disconnect(self) -> None:
        self._connected = False
        if self._listen_task:
            self._listen_task.cancel()
        if self._writer:
            try:
                await self._send_raw("QUIT :bye")
            except Exception:
                pass
            self._writer.close()
        logger.info("irc disconnected")

    async def receive(self) -> AsyncIterator[UnifiedMessage]:
        while self._connected:
            try:
                msg = await self._queue.get()
                yield msg
            except asyncio.CancelledError:
                break

    async def send(self, msg: OutboundMessage) -> str | None:
        target = msg.chat_id
        # Split long messages (IRC 512 byte limit)
        for line in msg.text.split("\n"):
            if line.strip():
                await self._send_raw(f"PRIVMSG {target} :{line}")
        self._last_activity = datetime.now()
        return None

    async def get_status(self) -> ChannelStatus:
        return ChannelStatus(connected=self._connected, channel="irc",
                             account_id=f"{self._nickname}@{self._server}",
                             last_activity=self._last_activity)

    async def _send_raw(self, line: str) -> None:
        if self._writer:
            self._writer.write((line + "\r\n").encode("utf-8"))
            await self._writer.drain()

    async def _listen(self) -> None:
        try:
            while self._connected and self._reader:
                raw = await self._reader.readline()
                if not raw:
                    break
                line = raw.decode("utf-8", errors="replace").strip()
                if line.startswith("PING"):
                    token = line.split(" ", 1)[1] if " " in line else ""
                    await self._send_raw(f"PONG {token}")
                    continue
                await self._process_line(line)
        except (asyncio.CancelledError, ConnectionError):
            self._connected = False

    async def _process_line(self, line: str) -> None:
        match = _PRIVMSG_RE.match(line)
        if not match:
            return

        nick, hostmask, target, text = match.groups()
        if nick == self._nickname:
            return

        self._last_activity = datetime.now()
        self._msg_counter += 1

        if text.startswith(self._prefix):
            parts = text[len(self._prefix):].split()
            mc = MessageContent(type=ContentType.COMMAND, text=text,
                                command=parts[0] if parts else "", args=parts[1:])
        else:
            mc = MessageContent(type=ContentType.TEXT, text=text)

        # For DMs, target is our nickname; reply to sender
        chat_id = target if target.startswith("#") else nick

        msg = UnifiedMessage(
            id=str(self._msg_counter), channel="irc",
            sender=Identity(id=nick, username=nick),
            content=mc, chat_id=chat_id, raw=line,
        )
        await self._queue.put(msg)
